pkg/controller/atlasdeployment/search_nodes.go

```diff
@@ -0,0 +1,312 @@
+	}
+
+	akoEmpty := len(s.deployment.Spec.DeploymentSpec.SearchNodes) == 0
+	atlasEmpty := !found
+
+	switch {
+	case !akoEmpty && atlasEmpty:
+		// If no nodes configured in atlas, but some in the operator - create them.
+		return s.create()
+	case !akoEmpty && !atlasEmpty:
+		// If nodes already configured in atlas and in the operator - update them.
+		return s.update(atlasNodes)
+	case akoEmpty && !atlasEmpty:
+		// If no nodes configured in the operator, but some in atlas - delete them.
+		return s.delete()
+	default:
+		// akoEmpty && atlasEmpty
+		// If no search nodes are in AKO and no nodes in Atlas - unmanage.
+		return s.unmanage()
+	}
+}
+
+// handleUpserting handles creating or updating search nodes. The following state is handled here:
+//
+//	Conditions:
+//	  Status:                False
+//	  Type:                  SearchNodesReady
+//	  Reason:                SearchNodesCreating | SearchNodesUpdating
+//
+// The following transitions can happen here:
+// - terminate: when an error occurred or when the operation has been aborted.
+// - idle: when search nodes are marked as IDLE in Atlas.
+func (s *searchNodeController) handleUpserting(state workflow.ConditionReason) workflow.Result {
+	if len(s.deployment.Spec.DeploymentSpec.SearchNodes) == 0 {
+		return s.terminate(workflow.ErrorSearchNodesOperationAborted, errors.New("aborting update/create: no search nodes specified"))
+	}
+
+	atlasNodes, found, err := s.getAtlasSearchDeployment()
+	switch {
+	case err != nil:
+		return s.terminate(workflow.ErrorSearchNodesNotUpsertedInAtlas, err)
+	case !found:
+		return s.terminate(workflow.ErrorSearchNodesNotUpsertedInAtlas, errors.New("no search nodes found in Atlas"))
+	case atlasNodes.GetStateName() != "IDLE":
+		return s.progress(
+			state,
+			fmt.Sprintf("search nodes are not ready yet, Atlas state: %q", atlasNodes.GetStateName()),
+			"waiting for search nodes to become ready",
+		)
+	default:
+		return s.idle()
+	}
+}
+
+// handleCreating handles creating search nodes. The following state is handled here:
+//
+//	Conditions:
+//	  Status:                False
+//	  Type:                  SearchNodesReady
+//	  Reason:                SearchNodesDeleting
+//
+// The following transitions can happen here:
+// - terminate: when an error occurred or when the operation has been aborted.
+// - unmanage: when there are no search nodes anymore in Atlas.
+func (s *searchNodeController) handleDeleting() workflow.Result {
+	if len(s.deployment.Spec.DeploymentSpec.SearchNodes) > 0 {
+		return s.terminate(workflow.ErrorSearchNodesOperationAborted, errors.New("aborting deletion: search nodes are specified"))
+	}
+
+	atlasNodes, found, err := s.getAtlasSearchDeployment()
+	if err != nil {
+		// transition back to pending, something went wrong.
+		return s.terminate(workflow.ErrorSearchNodesNotDeletedInAtlas, err)
+	}
+
+	switch {
+	case err != nil:
+		return s.terminate(workflow.ErrorSearchNodesNotUpsertedInAtlas, err)
```

Comment:
This if seems redundant:

```suggestion
	atlasNodes, found, err := s.getAtlasSearchDeployment()
	switch {
	case err != nil:
		return s.terminate(workflow.ErrorSearchNodesNotUpsertedInAtlas, err)
```